### ISSUE TITLE
Kernel damage if jupyter.display_mimetypes doesn’t overlap an object’s reprs

### ISSUE THREAD

**Issue by flying-sheep** [IRkernel/IRkernel#111]
i was dumb and accidentally set `jupyter.display_mimetypes` instead of `jupyter.plot_mimetypes` to only some image mimetypes.

this resulted in [this code](https://github.com/IRkernel/IRkernel/blob/cdf2d1f547b562f54e2ab4b265741decbbfd14fe/R/execution.r#L95-L105) sending an `execute_result` with `data` being a `list()` instead of a `namedlist()`.

JSON-conversion converted this into `[]` instead of `{}` which breaks the ipython server.

we should:

1. refactor [this](https://github.com/IRkernel/IRkernel/blob/cdf2d1f547b562f54e2ab4b265741decbbfd14fe/R/execution.r#L94-L105) and [this](https://github.com/IRkernel/IRdisplay/blob/master/R/display.R#L20-L28y) to delegate to a common function (as it’s the same code anyway)
2. use `namedlist()`
3. (optional) always append `text/plain` to `jupyter.display_mimetypes` so there’s always something to display

---

i filed this bug instead of doing it so that we can discuss #3

**Comment by JanSchulz** [comment]
I think 3 should happen, whatever happens with the rest.

**Comment by takluyver** [comment]
I see you implemented 2 in 74f7e7115e7fba210bb8cae02e23a1ad8acbe7ff and 1c357fdcf9a510d320fa495d8681ca300f2950e2 .

We appear to be somewhat ambiguous about whether a `text/plain` format is required. The messaging docs say it should always be there, but if you do `display_png()` in IPython, I don't think any text repr is sent. We certainly don't expect the dict to be empty though. So, extra alternatives to step 3:

- if the list of mimetypes to display is empty, display text/plain anyway.
- throw an R error when trying to display something with no formats specified.

I'm about even on all three of those options at the moment.

**Comment by JanSchulz** [comment]
but `display_png()` sends a `display_data` reply, not a `execute_result`.

just some random thoght: the "change option by specifying a list" is in this case error prone. It's much easier to "forget" a mimetype in a new (probably manually created) list which is set as this option (-> R) than when you have to select a formatter and disable it (-> python). Maybe a `disable_format(<mimetype(s)>)` and `enable_format(<mimetype(s)>)` as the "official" way to change the mimetype would prevent that. That way you can be sure that a missing `text/plain` really means that you don't want a text plain result.

I think such a convenience wrapper around the plot types would be nice, they are probably more used than the general types.

**Comment by takluyver** [comment]
True, but `display_data` and `execute_result` are mostly equivalent. I don't remember any discussion that text/plain is necessary for execute_result but not for display_data.

Configuring lists of things is always technically awkward - how do you combine different levels of config? Do later levels add to previous ones, or replace them? If they add, is there a way to subtract? What about adding something only if it's not already there? But there's quite a short list of viable mimetypes for plots - let's not overcomplicate the API. I think just setting the whole list is OK for now.

**Comment by janschulz** [comment]
What should happen here? Is this still relevant?
